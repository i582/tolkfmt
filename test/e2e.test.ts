import {join, normalize} from "path"
import {runCommand} from "./utils/test-util"
import {readFileSync, rmSync, writeFileSync} from "fs"
import {mkdir} from "fs/promises"

// disable tests on windows
const testExceptWindows = process.platform === "win32" && process.env["CI"] ? test.skip : test

const tolkfmt = (...args: string[]) => {
    const tolkfmtPath = normalize(join(__dirname, "..", "bin", "tolkfmt"))
    const command = `node ${tolkfmtPath} ${args.join(" ")}`
    return runCommand(command)
}

const outputDir = join(__dirname, "output")

const goodContract = `
fun test(x: int, y: string): string {
    return x + y
}
`

const badContract = `
fun   test(  x: int,y:string   ): string{
return    x+y;
}
`

const contractWithSyntaxError = `
fun test(x: int, y: string): string {
    return foo("hello world";
}
`

const structCode = `
struct MyStruct {
    x: int = 42,
    y: string = "hello"
}
`

describe("tolkfmt foo.tolk", () => {
    testExceptWindows("Exits with correct code", async () => {
        await mkdir(outputDir, {recursive: true})
        const file = join(outputDir, "contract.tolk")
        writeFileSync(file, goodContract)
        const result = await tolkfmt(file)
        expect(result).toMatchObject({kind: "exited", code: 0})

        rmSync(file)
    })

    testExceptWindows("Default run", async () => {
        await mkdir(outputDir, {recursive: true})
        const file = join(outputDir, "contract.tolk")
        writeFileSync(file, goodContract)
        const result = await tolkfmt(file)
        expect(result).toMatchSnapshot()

        rmSync(file)
    })

    testExceptWindows("Default run with write to file", async () => {
        await mkdir(outputDir, {recursive: true})
        const file = join(outputDir, "contract.tolk")
        writeFileSync(file, badContract)
        await tolkfmt(file, "-w")

        const newContent = readFileSync(file, "utf8")
        expect(newContent).toMatchSnapshot()

        rmSync(file)
    })

    testExceptWindows("Run on directory", async () => {
        const dir = outputDir
        const innerDir = join(dir, "inner")
        const innerInnerDir = join(innerDir, "inner-2")

        await mkdir(dir, {recursive: true})
        await mkdir(innerDir, {recursive: true})
        await mkdir(innerInnerDir, {recursive: true})

        // inner
        //   file1.tolk
        //   inner-2
        //      file2.tolk
        //      file3.tolk
        const file1 = join(innerDir, "file1.tolk")
        const file2 = join(innerInnerDir, "file2.tolk")
        const file3 = join(innerInnerDir, "file3.tolk")

        writeFileSync(file1, "fun foo1() {   }")
        writeFileSync(file2, "fun foo2() {  }")
        writeFileSync(file3, "fun foo3() {     }")

        await tolkfmt(innerDir, "-w")

        expect(readFileSync(file1, "utf8")).toMatchSnapshot()
        expect(readFileSync(file2, "utf8")).toMatchSnapshot()
        expect(readFileSync(file3, "utf8")).toMatchSnapshot()

        rmSync(innerDir, {recursive: true})
    })

    testExceptWindows("Check on directory with formatted files", async () => {
        const dir = outputDir
        const innerDir = join(dir, "inner")
        const innerInnerDir = join(innerDir, "inner-2")

        await mkdir(dir, {recursive: true})
        await mkdir(innerDir, {recursive: true})
        await mkdir(innerInnerDir, {recursive: true})

        const file1 = join(innerDir, "file1.tolk")
        const file2 = join(innerInnerDir, "file2.tolk")
        const file3 = join(innerInnerDir, "file3.tolk")

        writeFileSync(file1, "fun foo1() {}\n")
        writeFileSync(file2, "fun foo2() {}\n")
        writeFileSync(file3, "fun foo3() {}\n")

        const result = await tolkfmt(innerDir, "--check")
        expect(result).toMatchSnapshot()

        rmSync(innerDir, {recursive: true})
    })

    testExceptWindows("Check on directory with not formatted files", async () => {
        const dir = outputDir
        const innerDir = join(dir, "inner")
        const innerInnerDir = join(innerDir, "inner-2")

        await mkdir(dir, {recursive: true})
        await mkdir(innerDir, {recursive: true})
        await mkdir(innerInnerDir, {recursive: true})

        const file1 = join(innerDir, "file1.tolk")
        const file2 = join(innerInnerDir, "file2.tolk")
        const file3 = join(innerInnerDir, "file3.tolk")

        writeFileSync(file1, "fun foo1() {  }\n")
        writeFileSync(file2, "fun foo2() {  }\n")
        writeFileSync(file3, "fun foo3() {  }\n")

        const result = await tolkfmt(innerDir, "--check")
        expect(result).toMatchSnapshot()

        rmSync(innerDir, {recursive: true})
    })

    testExceptWindows("Check on several formatted files", async () => {
        const dir = outputDir
        const innerDir = join(dir, "inner")
        const innerInnerDir = join(innerDir, "inner-2")

        await mkdir(dir, {recursive: true})
        await mkdir(innerDir, {recursive: true})
        await mkdir(innerInnerDir, {recursive: true})

        const file1 = join(innerDir, "file1.tolk")
        const file2 = join(innerInnerDir, "file2.tolk")
        const file3 = join(innerInnerDir, "file3.tolk")

        writeFileSync(file1, "fun foo1() {}\n")
        writeFileSync(file2, "fun foo2() {  }\n") // not checked
        writeFileSync(file3, "fun foo3() {}\n")

        const result = await tolkfmt("--check", file1, file3)
        expect(result).toMatchSnapshot()

        rmSync(innerDir, {recursive: true})
    })

    testExceptWindows("Check on several directories", async () => {
        const dir = outputDir
        const innerDir = join(dir, "inner")
        const innerInnerDir = join(innerDir, "inner-2")
        const innerInnerDir2 = join(innerDir, "inner-3")

        await mkdir(dir, {recursive: true})
        await mkdir(innerDir, {recursive: true})
        await mkdir(innerInnerDir, {recursive: true})
        await mkdir(innerInnerDir2, {recursive: true})

        const file1 = join(innerDir, "file1.tolk")
        const file2 = join(innerInnerDir, "file2.tolk")
        const file3 = join(innerInnerDir2, "file3.tolk")

        writeFileSync(file1, "fun foo1() { }\n") // not checked
        writeFileSync(file2, "fun foo2() {  }\n")
        writeFileSync(file3, "fun foo3() {}\n")

        const result = await tolkfmt("--check", innerInnerDir, innerInnerDir2)
        expect(result).toMatchSnapshot()

        rmSync(innerDir, {recursive: true})
    })

    testExceptWindows("With syntax error", async () => {
        await mkdir(outputDir, {recursive: true})
        const file = join(outputDir, "contract.tolk")
        writeFileSync(file, contractWithSyntaxError)
        const result = await tolkfmt(file, "-w")

        // Check status without timing details
        expect(result.kind).toBe("exited")
        if (result.kind === "exited") {
            expect(result.code).toBe(0)
            expect(result.stderr).toBe("")
            expect(result.stdout).toMatch(/contract\.tolk \d+ms \(reformatted\)/)
        }

        rmSync(file)
    })

    testExceptWindows("Check and write flags simultaneously", async () => {
        await mkdir(outputDir, {recursive: true})
        const file = join(outputDir, "contract.tolk")
        writeFileSync(file, badContract)
        const result = await tolkfmt(file, "-w", "--check")
        expect(result).toMatchSnapshot()

        rmSync(file)
    })

    testExceptWindows("Format struct", async () => {
        await mkdir(outputDir, {recursive: true})
        const file = join(outputDir, "struct.tolk")
        writeFileSync(file, structCode)
        const result = await tolkfmt(file)
        expect(result).toMatchSnapshot()

        rmSync(file)
    })

    testExceptWindows("Version flag", async () => {
        const result = await tolkfmt("--version")
        expect(result).toMatchSnapshot()
    })

    testExceptWindows("Help flag", async () => {
        const result = await tolkfmt("--help")
        expect(result).toMatchSnapshot()
    })

    testExceptWindows("No arguments", async () => {
        const result = await tolkfmt()
        expect(result).toMatchSnapshot()
    })
})
