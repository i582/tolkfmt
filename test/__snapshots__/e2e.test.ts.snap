// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`tolkfmt foo.tolk Check and write flags simultaneously 1`] = `
{
  "code": 1,
  "kind": "exited",
  "stderr": "Error: Cannot use both --write and --check options together
",
  "stdout": "",
}
`;

exports[`tolkfmt foo.tolk Check on directory with formatted files 1`] = `
{
  "code": 0,
  "kind": "exited",
  "stderr": "",
  "stdout": "Checking formatting...
All Tolk files are properly formatted!
",
}
`;

exports[`tolkfmt foo.tolk Check on directory with not formatted files 1`] = `
{
  "code": 1,
  "kind": "exited",
  "stderr": "",
  "stdout": "Checking formatting...
[warn] file1.tolk
[warn] file3.tolk
[warn] file2.tolk
Code style issues found in the above files. Run tolkfmt with --write to fix.
",
}
`;

exports[`tolkfmt foo.tolk Check on several directories 1`] = `
{
  "code": 1,
  "kind": "exited",
  "stderr": "",
  "stdout": "Checking formatting...
[warn] file2.tolk
Code style issues found in the above files. Run tolkfmt with --write to fix.
",
}
`;

exports[`tolkfmt foo.tolk Check on several formatted files 1`] = `
{
  "code": 0,
  "kind": "exited",
  "stderr": "",
  "stdout": "Checking formatting...
All Tolk files are properly formatted!
",
}
`;

exports[`tolkfmt foo.tolk Default run 1`] = `
{
  "code": 0,
  "kind": "exited",
  "stderr": "",
  "stdout": "fun test(x: int, y: string): string {
    return x + y;
}
",
}
`;

exports[`tolkfmt foo.tolk Default run with write to file 1`] = `
"fun test(x: int, y: string): string {
    return x + y;
}
"
`;

exports[`tolkfmt foo.tolk Format struct 1`] = `
{
  "code": 0,
  "kind": "exited",
  "stderr": "",
  "stdout": "struct MyStruct {
    x: int = 42
    y: string = "hello"
}
",
}
`;

exports[`tolkfmt foo.tolk Help flag 1`] = `
{
  "code": 0,
  "kind": "exited",
  "stderr": "",
  "stdout": "tolkfmt/0.0.4

Usage:
  $ tolkfmt [options] <files or directories>

Options:
  -v, --version  Display version number 
  -w, --write    Write result to same file 
  -c, --check    Check if the given files are formatted 
  -h, --help     Display this message 
",
}
`;

exports[`tolkfmt foo.tolk No arguments 1`] = `
{
  "code": 0,
  "kind": "exited",
  "stderr": "",
  "stdout": "",
}
`;

exports[`tolkfmt foo.tolk Run on directory 1`] = `
"fun foo1() {}
"
`;

exports[`tolkfmt foo.tolk Run on directory 2`] = `
"fun foo2() {}
"
`;

exports[`tolkfmt foo.tolk Run on directory 3`] = `
"fun foo3() {}
"
`;

exports[`tolkfmt foo.tolk Version flag 1`] = `"tolkfmt/0.0.4"`;
